platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/consensus/AncestorIterator.java

```diff
@@ -0,0 +1,210 @@
+            // BOTTOM
+            mark.markVisited(curr); // mark this event, so we don't explore it again later
+            switch (state) {
+                case TRAVERSING_SELF_PARENT -> { // try to traverse into selfParent
+                    final EventImpl p = curr.getSelfParent();
+                    state = IteratorState.TRAVERSING_OTHER_PARENT;
+                    if (mark.isNotVisited(p) && valid.test(p)) {
+                        stackRef.push(curr);
+                        stackState.push(state);
+                        stackSelfAncestor.push(selfAncestor);
+                        stackTime.push(timeReachedRoot);
+                        curr = p;
+                        state = IteratorState.TRAVERSING_SELF_PARENT;
+                        if (selfAncestor) {
+                            timeReachedRoot = curr.getTimeCreated(); // ancestors of curr reached creator then
+                        }
+                    } // there is no selfParent, or it was already visited, or it was not valid
+                }
+                case TRAVERSING_OTHER_PARENT -> { // try to traverse into otherParent
+                    final EventImpl p = curr.getOtherParent();
+                    state = IteratorState.BOTTOM;
+                    if (mark.isNotVisited(p) && valid.test(p)) {
+                        stackRef.push(curr);
+                        stackState.push(state);
+                        stackSelfAncestor.push(selfAncestor);
+                        stackTime.push(timeReachedRoot);
+                        curr = p;
+                        state = IteratorState.TRAVERSING_SELF_PARENT;
+                        // first step off the selfAncestor path makes all the events below false
+                        selfAncestor = false;
+                    } // else there is no otherParent, or it was already visited, or it was not
+                    // valid
+                }
+                case BOTTOM -> { // done with ancestors of curr, so return curr then backtrack
+                    if (stackRef.isEmpty()) { // if we're back to the root
+                        hasNext = false; // then there are no more
+                        return curr; // return this root
+                    }
+                    final EventImpl toReturn = curr; // else we are done with all the descendents, so backtrack
+                    curr = stackRef.pop();
+                    state = stackState.pop();
+                    selfAncestor = stackSelfAncestor.pop();
+                    timeReachedRoot = stackTime.pop();
+                    return toReturn; // return the child of the vertex we just backtracked to
+                }
+                default -> throw new IllegalStateException("Unknown state: " + state);
+            }
+        }
+    }
+
+    private enum IteratorState {
+        TRAVERSING_SELF_PARENT,
+        TRAVERSING_OTHER_PARENT,
+        BOTTOM
+    }
```

Comment:
Nit, could this be moved to the top of the class? Normally enums come above methods.